File: app/src/main/java/com/vrem/wifianalyzer/wifi/accesspoint/AccessPointsAdapter.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer.wifi.accesspoint;

import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ExpandableListView;
import android.widget.ImageView;

import com.vrem.wifianalyzer.R;
import com.vrem.wifianalyzer.wifi.model.WiFiData;
import com.vrem.wifianalyzer.wifi.model.WiFiDetail;
import com.vrem.wifianalyzer.wifi.scanner.UpdateNotifier;

import androidx.annotation.NonNull;

class AccessPointsAdapter extends BaseExpandableListAdapter implements UpdateNotifier {
    private AccessPointsAdapterData accessPointsAdapterData;
    private AccessPointDetail accessPointDetail;
    private AccessPointPopup accessPointPopup;
    private ExpandableListView expandableListView;

    AccessPointsAdapter() {
        super();
        setAccessPointsAdapterData(new AccessPointsAdapterData());
        setAccessPointDetail(new AccessPointDetail());
        setAccessPointPopup(new AccessPointPopup());
        expandableListView = null;
    }

    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
        WiFiDetail wiFiDetail = (WiFiDetail) getGroup(groupPosition);
        View view = accessPointDetail.makeView(convertView, parent, wiFiDetail, false);
        attachPopup(view, wiFiDetail);

        ImageView groupIndicator = view.findViewById(R.id.groupIndicator);
        int childrenCount = getChildrenCount(groupPosition);
        if (childrenCount > 0) {
            groupIndicator.setVisibility(View.VISIBLE);
            groupIndicator.setImageResource(isExpanded
                ? R.drawable.ic_expand_less
                : R.drawable.ic_expand_more);
        } else {
            groupIndicator.setVisibility(View.GONE);
        }
        return view;
    }

    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        WiFiDetail wiFiDetail = (WiFiDetail) getChild(groupPosition, childPosition);
        View view = accessPointDetail.makeView(convertView, parent, wiFiDetail, true);
        attachPopup(view, wiFiDetail);
        view.findViewById(R.id.groupIndicator).setVisibility(View.GONE);
        return view;
    }

    @Override
    public void update(@NonNull WiFiData wiFiData) {
        accessPointsAdapterData.update(wiFiData, expandableListView);
        notifyDataSetChanged();
    }

    @Override
    public int getGroupCount() {
        return accessPointsAdapterData.parentsCount();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        return accessPointsAdapterData.childrenCount(groupPosition);
    }

    @Override
    public Object getGroup(int groupPosition) {
        return accessPointsAdapterData.parent(groupPosition);
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return accessPointsAdapterData.child(groupPosition, childPosition);
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return true;
    }

    @Override
    public void onGroupCollapsed(int groupPosition) {
        accessPointsAdapterData.onGroupCollapsed(groupPosition);
    }

    @Override
    public void onGroupExpanded(int groupPosition) {
        accessPointsAdapterData.onGroupExpanded(groupPosition);
    }

    void setAccessPointsAdapterData(@NonNull AccessPointsAdapterData accessPointsAdapterData) {
        this.accessPointsAdapterData = accessPointsAdapterData;
    }

    void setAccessPointDetail(@NonNull AccessPointDetail accessPointDetail) {
        this.accessPointDetail = accessPointDetail;
    }

    void setAccessPointPopup(@NonNull AccessPointPopup accessPointPopup) {
        this.accessPointPopup = accessPointPopup;
    }

    void setExpandableListView(ExpandableListView expandableListView) {
        this.expandableListView = expandableListView;
    }

    private void attachPopup(@NonNull View view, @NonNull WiFiDetail wiFiDetail) {
        View popupView = view.findViewById(R.id.attachPopup);
        if (popupView != null) {
            accessPointPopup.attach(popupView, wiFiDetail);
            accessPointPopup.attach(view.findViewById(R.id.ssid), wiFiDetail);
        }
    }

}

File: app/src/main/kotlin/com/vrem/wifianalyzer/wifi/accesspoint/AccessPointsAdapter.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2025 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer.wifi.accesspoint

import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.ExpandableListView
import android.widget.ImageView
import com.vrem.annotation.OpenClass
import com.vrem.wifianalyzer.R
import com.vrem.wifianalyzer.wifi.model.WiFiData
import com.vrem.wifianalyzer.wifi.model.WiFiDetail
import com.vrem.wifianalyzer.wifi.scanner.UpdateNotifier

@OpenClass
class AccessPointsAdapter(
    private val accessPointsAdapterData: AccessPointsAdapterData = AccessPointsAdapterData(),
    private val accessPointDetail: AccessPointDetail = AccessPointDetail(),
    private val accessPointPopup: AccessPointPopup = AccessPointPopup()
) : BaseExpandableListAdapter(), UpdateNotifier {

    lateinit var expandableListView: ExpandableListView

    override fun getGroupView(groupPosition: Int, expanded: Boolean, convertView: View?, parent: ViewGroup?): View {
        val wiFiDetail = getGroup(groupPosition)
        val view = accessPointDetail.makeView(convertView, parent, wiFiDetail)
        attachPopup(view, wiFiDetail)
        val groupIndicator = view.findViewById<ImageView>(R.id.groupIndicator)
        val childrenCount = getChildrenCount(groupPosition)
        if (childrenCount > 0) {
            groupIndicator.visibility = View.VISIBLE
            groupIndicator.setImageResource(if (expanded) R.drawable.ic_expand_less else R.drawable.ic_expand_more)
        } else {
            groupIndicator.visibility = View.GONE
        }
        return view
    }

    override fun getChildView(groupPosition: Int, childPosition: Int, lastChild: Boolean, convertView: View?, parent: ViewGroup?): View {
        val wiFiDetail = getChild(groupPosition, childPosition)
        val view = accessPointDetail.makeView(convertView, parent, wiFiDetail, true)
        attachPopup(view, wiFiDetail)
        view.findViewById<View>(R.id.groupIndicator).visibility = View.GONE
        return view
    }

    override fun update(wiFiData: WiFiData) {
        accessPointsAdapterData.update(wiFiData, expandableListView)
        notifyDataSetChanged()
    }

    override fun getGroupCount(): Int =
        accessPointsAdapterData.parentsCount()

    override fun getChildrenCount(groupPosition: Int): Int =
        accessPointsAdapterData.childrenCount(groupPosition)

    override fun getGroup(groupPosition: Int): WiFiDetail =
        accessPointsAdapterData.parent(groupPosition)

    override fun getChild(groupPosition: Int, childPosition: Int): WiFiDetail =
        accessPointsAdapterData.child(groupPosition, childPosition)

    override fun getGroupId(groupPosition: Int): Long = groupPosition.toLong()

    override fun getChildId(groupPosition: Int, childPosition: Int): Long = childPosition.toLong()

    override fun hasStableIds(): Boolean = true

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean = true

    override fun onGroupCollapsed(groupPosition: Int) =
        accessPointsAdapterData.onGroupCollapsed(groupPosition)

    override fun onGroupExpanded(groupPosition: Int) =
        accessPointsAdapterData.onGroupExpanded(groupPosition)

    private fun attachPopup(view: View, wiFiDetail: WiFiDetail) {
        view.findViewById<View>(R.id.attachPopup)?.let {
            accessPointPopup.attach(it, wiFiDetail)
            accessPointPopup.attach(view.findViewById(R.id.ssid), wiFiDetail)
        }
    }

}